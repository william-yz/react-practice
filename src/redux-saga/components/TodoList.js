import React from 'react'

import { Timeline } from 'antd'
import { connect } from 'react-redux'

import Todo from './Todo'
import Actions from './Actions'

import { getTodoList, saveTodo, removeTodo, toggleComplate } from '../actionCreators'
const style = {
  width: '300px'
}

class TodoList extends React.Component {
  addTodo = (todo) => {
    this.props.dispatch({
      type: 'ADD_TODO',
      payload: todo
    })
  }
  removeTodo = todo => () => {
    this.props.dispatch(removeTodo(todo))
  }
  save = todo => () => {
    this.props.dispatch(saveTodo(todo))
  }
  edit = todo => () => {
    this.props.dispatch({
      type: 'EDIT_TODO',
      payload: todo
    })
  }

  input = todo => (e) => {
    this.props.dispatch({
      type: 'INPUTTING_TODO',
      payload: {...todo, content: e.target.value}
    })
  }

  complate = todo => () => {
    this.props.dispatch(toggleComplate((todo)))
  }

  componentDidMount = () => {
    this.props.dispatch(getTodoList())
  }

  render () {
    return (
      <div>
        <Timeline>
          {this.props.todos.map((todo, index) => (
            <Timeline.Item
              key={todo.id}
              style={style}
              color={todo.complated ? 'green' : 'blue'}
              onDoubleClick={this.complate(todo)}
            >
              <Todo {...todo} otherIsEditting={this.props.someoneIsEditting} save={this.save(todo)} input={this.input(todo)} edit={this.edit(todo)} removeTodo={this.removeTodo(todo)}/>
            </Timeline.Item>
            ))}
        </Timeline>
        <Actions someoneIsEditting={this.props.someoneIsEditting}  addTodo={this.addTodo.bind(this,{
          content: '',
          editting: true
        })}/>
      </div>
    )
  }
}

const mapStateToProps = ({todos}) => {
  return {
    todos,
    someoneIsEditting: todos.some(todo => todo.editting)
  }
}

export default connect(mapStateToProps)(TodoList)