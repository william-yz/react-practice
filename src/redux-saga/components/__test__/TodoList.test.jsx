import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import { TodoListComponent as TodoList, mapStateToProps } from '../TodoList'

test('TodoList', () => {
  const dispatch = jest.fn()
  const todos = [{
    id: 'temp-1',
    content: 'test',
    editting: true,
    complated: false,
    deleted: false,
  }]
  const someoneIsEditting = true
  const component = renderer.create(
    <TodoList
      dispatch={dispatch}
      todos={todos}
      someoneIsEditting={someoneIsEditting}
    />)
  const todoList = component.toJSON()
  expect(todoList).toMatchSnapshot()
  expect(dispatch.mock.calls).toEqual([[{ type: 'LOAD_TODOS' }]])
})

test('TodoList 1', () => {
  const dispatch = jest.fn()
  const firstTodo = {
    id: 'temp-1',
    content: 'test',
    editting: false,
    complated: false,
    deleted: false,
  }
  const todos = [firstTodo, {
    id: 'temp-2',
    content: 'test',
    editting: true,
    complated: false,
    deleted: true,
  }, {
    id: 'temp-3',
    content: 'test',
    editting: false,
    complated: true,
    deleted: false,
  }]
  const someoneIsEditting = true
  const component = shallow(
    <TodoList
      dispatch={dispatch}
      todos={todos}
      someoneIsEditting={someoneIsEditting}
    />)
  const timelineItem = component.find('TimelineItem')
  expect(timelineItem.length).toBe(2)
  expect(timelineItem.first().prop('color')).toEqual('blue')
  timelineItem.first().prop('onDoubleClick')()
  expect(dispatch.mock.calls.pop()).toEqual([{
    type: 'TOGGLE_COMPLATE',
    payload: firstTodo }])
  expect(timelineItem.get(1).props.color).toEqual('green')
  const todo = component.find('Todo').first()

  todo.prop('save')()
  expect(dispatch.mock.calls.pop()).toEqual([{
    type: 'SAVE_TODO',
    payload: firstTodo }])

  todo.prop('input')({
    target: {
      value: '123',
    },
  })
  expect(dispatch.mock.calls.pop()).toEqual([{
    type: 'INPUTTING_TODO',
    payload: { ...firstTodo, content: '123' },
  }])

  todo.prop('edit')()
  expect(dispatch.mock.calls.pop()).toEqual([{
    type: 'EDIT_TODO',
    payload: firstTodo,
  }])

  todo.prop('removeTodo')()
  expect(dispatch.mock.calls.pop()).toEqual([{
    type: 'REMOVE_TODO',
    payload: firstTodo,
  }])

  const actions = component.find('Actions')
  expect(actions.prop('redoabled')).toBe(true)

  actions.prop('addTodo')()
  expect(dispatch.mock.calls.pop()).toEqual([{
    type: 'ADD_TODO',
    payload: {
      content: '',
      editting: true,
    },
  }])

  actions.prop('redo')()
  expect(dispatch.mock.calls.pop()).toEqual([{
    type: 'REDO_REMOVE',
  }])

  expect(mapStateToProps({
    todos,
  })).toEqual({
    todos,
    someoneIsEditting: true,
  })
})
