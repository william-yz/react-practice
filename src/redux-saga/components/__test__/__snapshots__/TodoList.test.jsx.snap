exports[`test TodoList 1`] = `
<div>
  <ul
    className="ant-timeline">
    <li
      className="ant-timeline-item ant-timeline-item-last"
      onDoubleClick={[Function]}
      style={
        Object {
          "width": "300px",
        }
      }>
      <div
        className="ant-timeline-item-tail" />
      <div
        className="ant-timeline-item-head ant-timeline-item-head-blue"
        style={
          Object {
            "borderColor": null,
          }
        } />
      <div
        className="ant-timeline-item-content">
        <div
          className="editable-cell">
          <div
            className="editable-cell-input-wrapper">
            <input
              className="ant-input"
              disabled={false}
              onChange={[Function]}
              onKeyDown={[Function]}
              type="text"
              value="test" />
            <i
              className="anticon anticon-check editable-cell-icon-check"
              onClick={[Function]} />
          </div>
        </div>
      </div>
    </li>
  </ul>
  <span>
    <button
      className="ant-btn ant-btn-primary"
      disabled={true}
      onClick={[Function]}
      onMouseUp={[Function]}
      type="button">
      <span>
        Add
      </span>
    </button>
    Â 
    <button
      className="ant-btn ant-btn-primary"
      disabled={true}
      onClick={[Function]}
      onMouseUp={[Function]}
      type="button">
      <span>
        Redo
      </span>
    </button>
  </span>
</div>
`;
